src/pytorch_lightning/plugins/io/torch_plugin.py

```diff
@@ -13,21 +13,52 @@
 # limitations under the License.
 import logging
 import os
+import queue
 from typing import Any, Callable, Dict, Optional
 
 import pytorch_lightning as pl
 from pytorch_lightning.plugins.io.checkpoint_plugin import CheckpointIO
-from pytorch_lightning.utilities.cloud_io import atomic_save, get_filesystem
+from pytorch_lightning.utilities.cloud_io import _atomic_save, atomic_save, get_filesystem
 from pytorch_lightning.utilities.cloud_io import load as pl_load
-from pytorch_lightning.utilities.rank_zero import rank_zero_warn
+from pytorch_lightning.utilities.cloud_io import ThreadQueue
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.rank_zero import rank_zero_info, rank_zero_warn
 from pytorch_lightning.utilities.types import _PATH
 
 log = logging.getLogger(__name__)
 
 
 class TorchCheckpointIO(CheckpointIO):
     """CheckpointIO that utilizes :func:`torch.save` and :func:`torch.load` to save and load checkpoints
-    respectively, common for most use cases."""
+    respectively, common for most use cases.
+
+    Args:
+        save_async: whether to save the checkpoint asynchronously or not.
+        num_threads: Number of threads to use for asynchronous checkpointing.
+    """
+
+    def __init__(self, save_async: bool = False, num_threads: Optional[int] = None):
```

Comment:
I'm not convinced that these should be flags on this plugin. IMO it should be a separate `AsyncTorchCheckpointIO` class. This will be very useful for code that does not support the async mechanism, because then it can check `if isisntance(AsyncTorchCheckpointIO)`